service/bitbucket/annotator.go

```diff
@@ -0,0 +1,198 @@
+	}
+
+	return &ReportAnnotator{
+		cli:         cli,
+		sha:         sha,
+		owner:       owner,
+		repo:        repo,
+		annotations: annotations,
+		severityMap: map[rdf.Severity]string{
+			rdf.Severity_INFO:    annotationSeverityLow,
+			rdf.Severity_WARNING: annotationSeverityMedium,
+			rdf.Severity_ERROR:   annotationSeverityHigh,
+		},
+		duplicates: map[string]struct{}{},
+	}
+}
+
+// Post accepts a comment and holds it. Flush method actually posts comments to
+// Bitbucket in batch.
+func (r *ReportAnnotator) Post(_ context.Context, c *reviewdog.Comment) error {
+	c.Result.Diagnostic.GetLocation().Path = filepath.ToSlash(
+		filepath.Join(r.wd, c.Result.Diagnostic.GetLocation().GetPath()))
+	r.muAnnotations.Lock()
+	defer r.muAnnotations.Unlock()
+
+	anot := r.annotationFromReviewDogComment(*c)
+
+	// deduplicate event, because some reporters might report
+	// it twice, and bitbucket api will complain on duplicated
+	// external id of annotation
+	if _, ok := r.duplicates[*anot.ExternalId]; !ok {
+		r.annotations[c.ToolName] = append(r.annotations[c.ToolName], anot)
+	}
+
+	r.duplicates[*anot.ExternalId] = struct{}{}
+	return nil
+}
+
+// Flush posts comments which has not been posted yet.
+func (r *ReportAnnotator) Flush(ctx context.Context) error {
+	r.muAnnotations.Lock()
+	defer r.muAnnotations.Unlock()
+
+	// create/update/annotate report per tool
+	for tool, annotations := range r.annotations {
+		reportID := reportID(reporter, tool)
+		title := fmt.Sprintf("[%s] %s report", tool, reporter)
+		if len(annotations) == 0 {
+			// if no annotation, create Passed report
+			if err := r.createOrUpdateReport(ctx, reportID, title, reportResultPassed); err != nil {
+				return err
+			}
+			// and move one
+			continue
+		}
+
+		// create report or update report first, with the failed status
+		if err := r.createOrUpdateReport(ctx, reportID, title, reportResultFailed); err != nil {
+			return err
+		}
+
+		// send annotations in batches, because of the api max payload size limit
+		for start, annCount := 0, len(annotations); start < annCount; start += annotationsBatchSize {
+			end := start + annotationsBatchSize
+
+			if end > annCount {
+				end = annCount
+			}
+
+			// add annotations to the report
+			_, resp, err := r.cli.ReportsApi.BulkCreateOrUpdateAnnotations(
+				ctx, r.owner, r.repo, r.sha, reportID,
+			).Body(annotations[start:end]).Execute()
+
+			if err := checkAPIError(err, resp, http.StatusOK); err != nil {
+				return fmt.Errorf("bitbucket.BulkCreateOrUpdateAnnotations: %s", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *ReportAnnotator) annotationFromReviewDogComment(c reviewdog.Comment) openapi.ReportAnnotation {
+	a := openapi.NewReportAnnotation()
+	switch c.ToolName {
+	// TODO: different type of annotation based on tool?
+	default:
+		a.SetAnnotationType(annotationTypeCodeSmell)
+	}
+
+	// hash the output of linter and use it as external id
+	a.SetExternalId(hashString(c.Result.Diagnostic.OriginalOutput))
```

Comment:
OriginalOutput is actually not first-class field of the Diagnostic message and we cannot always assume the original output contains the message along with the location. (e.g. eslint stylish format).

Can you use the hash of Diagnostic proto message instead like this? 

```go
func hash(b []byte) string {
	h := sha256.New()
	_, _ = h.Write(b)
	return fmt.Sprintf("%x", h.Sum(nil))
}

// Note that it might be good to create external ID from the diagnostic
// content along with *original line* (by using git blame for example) to
// generate unique ID, but it hashes the Diagnostic message for simplicity.
func externalIDFromDiagnostic(d *rdf.Diagnostic) string {
	b, err := proto.Marshal(d)
	if err != nil {
		b = []byte(d.OriginalOutput)
	}
	return hash(b)
}
```